---
  title: "Class_exercise"
author: "B Fortunato"
date: "3/16/2023"
output: github_document
---
  
  
  # Load the libraries you need
  # Load functions you need "my_class_functions"
```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(ggplot2)
library(tidyverse)
library(IRanges)
source("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/util/my_class_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/util/my_class_functions2.R")
```



# load in your peak files for each replicate of each protein
# Here I am starting to analyze my data for my proteins of interest:
# proteinX, Y, Z 
# First I will read in each replicate file
```{r load in peak files}
# setting file paths
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/chipseqprofs"
peak_path <- "results/bwa/mergedLibrary/macs/broadPeak_good_files/"
broadpeakfilepath <- file.path(basepath, peak_path)



# printing out a table of the number of peaks in each file:
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)
peak_list

peak_num <- sapply(peak_list, length) %>% as.data.frame(row.names = T)
names(peak_num) <- c("num_peaks")
peak_num <- peak_num %>%
  rownames_to_column(var = "dbp") %>%
  separate(col = dbp,  into = c('dbp', 'replicate'), sep = "_")
peak_num[6,]
peak_num[7,]
peak_num[8,]
peak_num[9,]

```


# Now I am going to create consensus peaks for each protein
```{r consensus peaks}

dbp <- c("JUN_R1","JUN_R2","JUN_R3","JUN_R4")


# now run our function consensus_from_reduced
consensus_list <- lapply(dbp, consensus_from_reduced, peak_list)
names(consensus_list) <- dbp



num_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")
consensus_list


basepath <- "/scratch/Shares/rinnclass/CLASS_2023/befo3150"
consensus_path <- "CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/"
exportpath <- file.path(basepath, consensus_path)

for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}

peak_1 <- read_tsv("00_consensus_peaks/JUN_R1_consensus_peaks.bed", col_names = F)
peak_2 <- read_tsv("00_consensus_peaks/JUN_R2_consensus_peaks.bed", col_names = F)
peak_3 <- read_tsv("00_consensus_peaks/JUN_R3_consensus_peaks.bed", col_names = F)
peak_4 <- read_tsv("00_consensus_peaks/JUN_R4_consensus_peaks.bed", col_names = F)


```

# Now I am going to make my consensus peaks compatable with UCSC genome browser
```{r}
consensus_file_list <- list.files("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks", full.names = T, pattern = ".bed")

peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))

names(peaks) <- dbp

canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

# let's use lapply with filter funciton to cannonical_chr
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))


# now that these are all nice and clean let's export:
new_filenames <- paste0("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/", names(peaks), "_consensus.bed")


for(i in 1:length(peaks)) {
rtracklayer::export(peaks[[i]], paste0("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/", names(peaks)[i], "_consensus.bed") )}

for(i in 1:length(peaks)) {
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

headers <- paste0("track type=bed name=", names(peaks))
headers

# creating a path to export after we add header in for loop below
new_filenames <- paste0("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/ucsc_consensus_peaks/", names(peaks), ".bed")
new_filenames

for(i in 1:length(peaks)) {
rtracklayer::export(peaks[[i]], paste0("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/ucsc_consensus_peaks/", names(peaks)[i], ".bed") )}

for(i in 1:length(peaks)) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}


```

# I am curious if my proteins are transcription factors so I will use the annotations
# in a cell paper I found and see

```{r}
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/befo3150"
peak_path <- "CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks"
consensusPeakPath <- file.path(basepath, peak_path)
consensus_peaks_files <- list.files(consensusPeakPath, 
                                             pattern = "*.bed",
                                             full.names = TRUE)
consensus_peaks <- lapply(consensus_peaks_files, rtracklayer::import)
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/|_consensus.bed","", consensus_peaks_files)
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

# now just type == gene annotations
gencode_genes <- gencode_gr[gencode_gr$type == "gene"] 
table(gencode_gr$type)
rtracklayer::export(gencode_genes, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/gencode_genes.gtf")

# mRNA genes (called "protein_coding") in this version of gencode changes sometimes !
mrna_genes <- gencode_genes[gencode_genes$gene_type %in% "protein_coding"] 

rtracklayer::export(mrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/mrna_genes.gtf")
table(gencode_genes$gene_type)

# now doing a second index for lncRNA:
lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% "lncRNA"] 

rtracklayer::export(lncrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/lncrna_genes.gtf")
mrna_lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% c("protein_coding","lncRNA")]
rtracklayer::export(mrna_lncrna_genes, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/mrna_lncrna_genes.gtf")
lncrna_mrna_genes <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/mrna_lncrna_genes.gtf")

# Nice that loaded so much faster -- lets see as DF
lncrna_mrna_genes_df <- lncrna_mrna_genes %>% as.data.frame()

lncrna_mrna_promoters <- promoters(lncrna_mrna_genes, upstream = 1000, downstream = 1000)
# check right size
width(lncrna_mrna_promoters)
rtracklayer::export(lncrna_mrna_promoters, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/gene_annotations/lncrna_mrna_promoters.gtf")



lncrna_gene_ids <- mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "lncRNA"]
table(mrna_lncrna_genes$gene_type)
# same for mRNAs
mrna_gene_ids <-mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "protein_coding"]


num_peaks_df <- data.frame("dbp" = names(consensus_peaks),
                           "num_peaks" = sapply(consensus_peaks, length))


# Now let's get the total amount of the genome covered by all the peaks for a given DBP.

num_peaks_df$total_peak_length <- sapply(consensus_peaks, function(x) sum(width(x)))

promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, type = "counts")

# ok we see this is a large matrix of the number of overlaps at each promoter
# cols are promoters, rows are DBPS
# so now we just row sum for each DBP !

num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# Now let's break these promoters into two groups "lncrna" and "mrna"
# We will use the gene_id objects we made above to index and separate them.

num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

# mrna promoter overlaps
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])

# Let's take a look.
write_csv(num_peaks_df, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/num_peaks_df.csv")

```



genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                consensus_peaks, 
                                                type = "counts")

# Now let's extract the overlaps the same way we did for promoters above

# All gene bodies
num_peaks_df$peaks_overlapping_genebody <- 
  rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- 
  rowSums(genebody_peak_counts[,mrna_gene_ids])

# let's take a look.
# cool lots of info let's save:
write_csv(num_peaks_df, "num_peaks_df.csv")


url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"

destination_for_url <- "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/TF_annotations.xlsx"

# to download we can use download.file
download.file(url, destination_for_url)
human_tfs <- readxl::read_excel("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)


# let's rename the 4th column to indicate if it is a TF.
names(human_tfs)[4] <- "is_tf"

# now let's intersect gene names that are in our ChIP data and has TF identity.
length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))
human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]


# adding new column names
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

# Now comes the actual merge. Let's look at it first
# ?merge

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values.
dim(num_peaks_df[is.na(num_peaks_df$tf),])

# Ok looks good let's write it out:
num_peaks_df <- num_peaks_df[,1:12]
write_csv(num_peaks_df, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/num_peaks_df.csv")




# if you leave the object name you just created in the environment
# it will print out in the knit. For example :
num_peaks_df[,1:12]

```




# Now I want to compare a protein with a previous analysis 
```{r}

# goto UCSC genome browser and load in a peak file for a given protein
# load in the data for the same protein from the previous analysis
# compare how your consensus peaks are similar or different to previous analyses


url2 <- "https://bchm5631sp2021.s3-us-west-2.amazonaws.com/peaks/JUN_R1.bb"

destination_for_url <- "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/TF2_annotations.xlsx"

download.file(url, destination_for_url)

human_tfs2 <- readxl::read_excel("/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/TF2_annotations.xlsx",
                                sheet = 2, skip = 1)

names(human_tfs2)[4] <- "is_tf"

length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs2$Name)))
human_tfs2 <- human_tfs2[tolower(human_tfs2$Name) %in% tolower(num_peaks_df$dbp), 1:4]


names(human_tfs2) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

num_peaks_df2 <- merge(num_peaks_df, human_tfs2, all.x = T)

dim(num_peaks_df2[is.na(num_peaks_df2$tf),])

num_peaks_df2 <- num_peaks_df2[,1:12]
write_csv(num_peaks_df2, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/num_peaks_df2.csv")




num_peaks_df[,1:12]
num_peaks_df2[,1:12]


```


# Now I am going to determine how my peaks for each protein overlap annotations of the genome
# First I will find the overlaps between my consensus peaks with promoters of lncRNA and mRNA promoters

```{r}


# find overlaps of promoters for each protein


num_peaks_df2$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)
num_peaks_df2$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

# mrna promoter overlaps
num_peaks_df2$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])

# Let's take a look.
write_csv(num_peaks_df2, "/scratch/Shares/rinnclass/CLASS_2023/befo3150/CLASS_2023/CLASSES/05_R_analyses/00_consensus_peaks/consensus_peaks/results/num_peaks_df2.csv")
num_peaks_df
num_peaks_df2

```

## results: 
#1) What can you determine from these overlaps?
Given that these returned the same overlaps and peks numbers it's accurate to use the data we ran as it is equal in what we can learn.


## results:
# 1) What is the difference in overlaps between mRNA and lncRNA promoters
There weren't  differences in the overlaps for mRNA and lncRNA promoters.

# Now I am going to test if there is more binding over gene bodies than promoters
# I will seperate lncRNA and mRNA gene bodies to find the overlaps 

```{r}



```
## results: 
# 1) Do my proteins have more overlaps with promoters or genebodies?



# It is nice and all to find overlaps, but I am interested in how many proteins
# bind a specific promoter. I will use my handy "occurence" parameter in 
# " count peaks per feature" 

```{r}
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, 
                                               type = "occurrence")
promoter_peak_occurence

```
## results: I find the max number of proteins on a promoter to be X


# Now I want to start plotting my results
# First I will see if there is a realtionship between peak number and total DNA covered
```{r}


ggplot ()
```

# Now I want to color my plot by wether the protein is a TF or not.
```{r}


ggplot
```

# I want to make a histogram of the number of peaks for each of my proteins

```{r}



hist
```


# Now I want to facet this by the type of DNA binding domain my protein has.
```{r}

```


# Cool now I am ready to send my result to my collaborator as a
# Knitted document

